internal/servicelog/ringbuffer.go

```diff
@@ -0,0 +1,369 @@
+	rb := RingBuffer{
+		data: make([]byte, size),
+	}
+	return &rb
+}
+
+// Close closes the writer to further writes, readers may continue.
+func (rb *RingBuffer) Close() error {
+	rb.rwlock.Lock()
+	defer rb.rwlock.Unlock()
+	if rb.writeClosed {
+		return nil
+	}
+	rb.writeClosed = true
+	rb.releaseIterators()
+	return nil
+}
+
+// Closed returns true if the writing side has closed.
+func (rb *RingBuffer) Closed() bool {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.writeClosed
+}
+
+// Write writes p to the backing buffer, allocating the number of bytes in p.
+// If p is larger than the size of the buffer then io.ErrShortWrite is returned and the
+// number of bytes written. If the p is larger than the number of bytes available,
+// then the tail is discarded to make room.
+func (rb *RingBuffer) Write(p []byte) (written int, _ error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+	defer func() {
+		if written > 0 {
+			rb.signalIterators()
+		}
+	}()
+	rb.rwlock.Lock()
+	defer rb.rwlock.Unlock()
+	if rb.writeClosed {
+		return 0, io.ErrClosedPipe
+	}
+	size := rb.Size()
+	writeLength := len(p)
+	if writeLength > size {
+		writeLength = size
+	}
+	available := rb.available()
+	if available < writeLength {
+		err := rb.discard(writeLength - available)
+		if err != nil {
+			return 0, err
+		}
+	}
+	start := rb.writeIndex
+	end := rb.writeIndex + RingPos(writeLength)
+	low := int(start % RingPos(len(rb.data)))
+	high := int(end % RingPos(len(rb.data)))
+	if high == 0 {
+		high = len(rb.data)
+	}
+	if low < high {
+		copy(rb.data[low:high], p[0:writeLength])
+	} else {
+		lowLength := len(rb.data) - low
+		copy(rb.data[low:], p)
+		copy(rb.data[:high], p[lowLength:])
+	}
+	rb.writeIndex += RingPos(writeLength)
+	if writeLength < len(p) {
+		return writeLength, io.ErrShortWrite
+	}
+	return writeLength, nil
+}
+
+// Available returns the number of bytes available to allocate.
+func (rb *RingBuffer) Available() int {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.available()
+}
+
+func (rb *RingBuffer) available() int {
+	return len(rb.data) - int(rb.writeIndex-rb.readIndex)
+}
+
+// Buffered returns the number of bytes readable from the buffer.
+func (rb *RingBuffer) Buffered() int {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.buffered()
+}
+
+func (rb *RingBuffer) buffered() int {
+	return int(rb.writeIndex - rb.readIndex)
+}
+
+// Size returns the size in bytes of the internal buffer.
+func (rb *RingBuffer) Size() int {
+	return len(rb.data)
+}
+
+// Positions returns the start and end positions of readable data in the RingBuffer.
+func (rb *RingBuffer) Positions() (start RingPos, end RingPos) {
+	rb.rwlock.RLock()
+	defer rb.rwlock.RUnlock()
+	return rb.readIndex, rb.writeIndex
+}
+
+// Copy copies bytes from the range into the supplied dest buffer. If dest is not large enough
+// to fill the bytes from start to end, then start to start+len(dest) is copied and
+// the error io.ErrShortBuffer is returned.
```

Comment:
The error condition may be unexpected eventually, as in general filling the space available in dest is the success case of copy functions in Go. Probably not a big deal given that this is an internal API though. Just in case, though: are we making use of this error, or would using the returned `n` as usual with `copy` make the logic just as simple for call sites?

Reply:
ahhh this is the old comment, I need to fix this sorry. Copy no longer takes an end position, therefore ErrShortBuffer is no longer a possible error state.